### PR TITLE
Make Cromwell more resilient to GCS IOException

### PR DESCRIPTION
**Pull request by myazinn**
The description of the problem can be found here
https://broadworkbench.atlassian.net/browse/BA-5881
Unfortunately, there is no direct way to reproduce `500 Internal Server Error Backend Error`. Therefore, it is unclear how to show that this PR solves the problem.
However, since this error causes IOException, we think that reproducing some other IOException with this file should reproduce the problem close enough. If you are interested, we have an [experimental branch](https://github.com/EpamLifeSciencesTeam/cromwell/pull/9) for reproducing IOException. In that branch, Cromwell stops during execution, giving us time to delete the `rc` file if we want to. It allows us to see that this fix actually works and Cromwell indeed tries to read the file multiple times.

### CODE REVIEW

**Comment by myazinn** [comment]
Hi @cjllanwarne 
The `IOException` case was added in the PR #4272. It looks like it was done to solve a specific problem.
In fact, we are doing the same thing now. We've been tempted to make all `IOExceptions` retryable. However, we assumed that checking only specific exceptions were made on purpose. Therefore we decided to stick to this idea.
You know Cromwell better than us. If you think that this is OK to retry all `IOExceptions`, we will implement this.

**Comment by cjllanwarne** [comment]
My gut feeling: Let's leave it at this for now - but if we end up having to keep come back and add new flavors of `IOException` to this list maybe then we can just make it a global retry on `IOException`

**Comment by cjllanwarne** [comment]
UPDATE: If we can make this logic retry all 500 errors but leave any other new `IOException`s un-retried I think that's the best way to go for now. Does that sounds feasible? Thanks!

**Comment by myazinn** [comment]
Hi @cjllanwarne 
I'm not sure that I'm fully understood what you are wanting.

> leave any other new IOExceptions un-retried

Do you mean that we should leave the existing case from PR #4272 as is? Instead of changing it, we should add a case for any throwable to check whether it contains "500 Internal Server Error", right?

**Comment by aednichols** [comment]
Is it possible to add `500` to `AdditionalRetryableHttpCodes`?

**Comment by aednichols** [comment]
It's also possible that by
>all 500 errors

@cjllanwarne means 5xx

**Comment by myazinn** [comment]
Hi @aednichols !

> Is it possible to add `500` to `AdditionalRetryableHttpCodes`?

It is possible, but it won't give the result we need.
These codes are used only for `StorageException`s, since other exceptions don't have `getCode` method. Therefore, if we add `500` to `AdditionalRetryableHttpCodes`, Cromwell won't retry IOException caused by `500 Internal Server Error`.

> It's also possible that by all 500 errors @cjllanwarne means 5xx

We did not think about it :) I think you're right, but just in case we will wait for an answer.

**Comment by aednichols** [comment]
The values in `AdditionalRetryableHttpCodes` are evaluated against `gcs.getCode` not `gcs.getMessage`, so the fact the error copy is different shouldn't matter (so I _think_ it should work for you)

**Comment by cjllanwarne** [comment]
Yeah, that's right, I meant "any 5xx could be retried" seems like a good idea to me

**Comment by myazinn** [comment]
Hi @cjllanwarne :)
I've made some changes and now any 5xx exception is retryable. Can you check please if this solution is suitable?

**Comment by myazinn** [comment]
I had to reopen the PR since one test failed due to external reasons and I couldn't find a button to restart a single test.

**Comment by gsaksena** [comment]
Just curious whether the retry has (or ought to have) an exponential
backoff.  500's may be triggered by heavy bucket reads increasing faster
than the bucket can scale.

File: engine/src/test/scala/cromwell/engine/io/IoActorSpec.scala [hunk 1]
```diff
@@ -221,10 +222,26 @@ class IoActorSpec extends TestKitSuite with FlatSpecLike with Matchers with Impl
...
     )
 
     retryables foreach { IoActor.isRetryable(_) shouldBe true }
     retryables foreach { IoActor.isFatal(_) shouldBe false }
   }
+
+  it should "have correct non-retryable exceptions" in {
+    val nonRetryables = List(
+      new IOException("404 File Not Found"),
+      new Exception("5xx HTTP Status Code")
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/io/IoActorSpec.scala]: I think your regex will correctly not identify this, but can we make sure with one more test case?

`new IOException("Could not read from gs://fc-secure-<snip>/JointGenotyping/<snip>/call-HardFilterAndMakeSitesOnlyVcf/shard-500/rc: 404 File Not Found")`

**Comment by cjllanwarne** [comment]
@gsaksena correct - the backoff is defined here in case your interested in the details: https://github.com/broadinstitute/cromwell/blob/develop/core/src/main/scala/cromwell/core/io/IoCommand.scala#L13

**Comment by gsaksena** [comment]
Cool!

File: engine/src/test/scala/cromwell/engine/io/IoActorSpec.scala (hunk 1 above)
Comment (myazinn) [reply engine/src/test/scala/cromwell/engine/io/IoActorSpec.scala]: Good point. I added a new test case for it.

**Comment by aednichols** [comment]
Hi @myazinn, unfortunately we had to revert this PR because it causes our CI to hang forever and time out.

It was our mistake that we did not properly run tests on the branch before merging.

We are working on fixing this so external contributors get the same test coverage as internal PRs.

**Comment by myazinn** [comment]
@aednichols
It’s strange that it happened. 
Do you have any idea why this PR caused such a problem?
As far as I know, Cromwell shouldn't have been to retry 5xx exception forever. So I'm curious what could have gone wrong.

**Pull request merged**